src/targets/gpu/kernels/include/migraphx/kernels/gather.hpp

```diff
@@ -0,0 +1,57 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015-2022 Advanced Micro Devices, Inc. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#ifndef MIGRAPHX_GUARD_KERNELS_GATHER_HPP
+#define MIGRAPHX_GUARD_KERNELS_GATHER_HPP
+
+#include <migraphx/kernels/index.hpp>
+#include <migraphx/kernels/shape.hpp>
+#include <migraphx/kernels/algorithm.hpp>
+
+namespace migraphx {
+
+template <int Axis, class T, class U, class V>
+__device__ void gather(const T& data_t, const U& indices_t, const V& output_t)
+{
+    auto ind           = make_index();
+    auto lengths       = data_t.get_shape().lens;
+    auto axis_dim_size = lengths[Axis];
+
+    lengths[Axis] = indices_t.get_shape().elements();
+
+    auto out_comp = make_shape(lengths, output_t.get_shape().strides);
```

Comment:
This needs to be a constexpr variable: `constexpr auto out_comp = ...`, but I think you will need to extract this to a seperate `constexpr` function to do this:

```cpp
template<int Axis, class T, class U>
constexpr auto gather_shape(T input, U indices)
{
    auto lengths       = input.get_shape().lens;
    auto axis_dim_size = lengths[Axis];

    lengths[Axis] = indices.get_shape().elements();

    return make_shape(lengths, input.get_shape().strides);
}
```

It would also be better to use the strides from `data_t`, since thats where lens are coming from as well. We dont use strides either way.

Reply:
Ah I see, I just did what the previous device function used and got that working. I can change and retest

Reply:
done